File: android-native/app/src/main/java/in/videh/filedtracker/bglocation/TrackingApi.java
package in.videh.filedtracker.bglocation;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public final class TrackingApi {
    private static final String TAG = "FTTrackingApi";
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
    private static final ExecutorService IO = Executors.newFixedThreadPool(2);
    private static final OkHttpClient HTTP = in.videh.filedtracker.nativeapp.AppConfig.okHttpBuilder()
            .connectTimeout(12, TimeUnit.SECONDS)
            .readTimeout(20, TimeUnit.SECONDS)
            .writeTimeout(20, TimeUnit.SECONDS)
            .callTimeout(25, TimeUnit.SECONDS)
            .retryOnConnectionFailure(true)
            .build();

    private static Request.Builder req(String apiBase, String token, String path) {
        return new Request.Builder()
                .url(apiBase + path)
                .addHeader("Authorization", "Bearer " + token)
                .addHeader("X-Client-Source", "native")
                .addHeader("Content-Type", "application/json");
    }

    public static boolean postIntervalSnapshot(String apiBase, String token, int slot, double lat, double lng) {
        AtomicBoolean ok = new AtomicBoolean(false);
        try {
            java.util.concurrent.Future<?> f = IO.submit(() -> {
                try {
                    JSONObject body = new JSONObject();
                    body.put("slot", slot);
                    body.put("lat", lat);
                    body.put("lng", lng);
                    Request request = req(apiBase, token, "/api/attendance/interval-snapshot")
                            .post(RequestBody.create(body.toString(), JSON))
                            .build();
                    try (Response res = HTTP.newCall(request).execute()) {
                        ok.set(res.isSuccessful());
                        if (!res.isSuccessful()) {
                            String msg = res.body() != null ? res.body().string() : "";
                            Log.w(TAG, "interval-snapshot " + res.code() + " " + msg);
                        }
                    }
                } catch (Exception e) {
                    Log.e(TAG, "postIntervalSnapshot failed", e);
                }
            });
            f.get(20, TimeUnit.SECONDS);
        } catch (Exception e) {
            Log.e(TAG, "postIntervalSnapshot wait failed", e);
        }
        return ok.get();
    }

    public static void postHeartbeat(String apiBase, String token, double lat, double lng) {
        IO.execute(() -> {
            try {
                JSONObject hb = new JSONObject();
                hb.put("lat", lat);
                hb.put("lng", lng);
                JSONObject body = new JSONObject();
                body.put("points", new JSONArray());
                body.put("heartbeat", hb);
                Request request = req(apiBase, token, "/api/attendance/track")
                        .post(RequestBody.create(body.toString(), JSON))
                        .build();
                try (Response res = HTTP.newCall(request).execute()) {
                    if (!res.isSuccessful()) Log.w(TAG, "track heartbeat " + res.code());
                }
            } catch (Exception e) {
                Log.w(TAG, "postHeartbeat failed", e);
            }
        });
    }

    public static void postTrackBatch(
            String apiBase,
            String token,
            JSONArray points,
            JSONArray mapProbes,
            double mapSpreadM,
            double hbLat,
            double hbLng
    ) {
        IO.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("points", points != null ? points : new JSONArray());
                if (mapProbes != null && mapProbes.length() > 0) body.put("mapProbes", mapProbes);
                if (mapSpreadM > 0) body.put("mapGpsSpreadM", mapSpreadM);
                JSONObject hb = new JSONObject();
                hb.put("lat", hbLat);
                hb.put("lng", hbLng);
                body.put("heartbeat", hb);
                Request request = req(apiBase, token, "/api/attendance/track")
                        .post(RequestBody.create(body.toString(), JSON))
                        .build();
                try (Response res = HTTP.newCall(request).execute()) {
                    if (!res.isSuccessful()) Log.w(TAG, "track batch " + res.code());
                }
            } catch (Exception e) {
                Log.w(TAG, "postTrackBatch failed", e);
            }
        });
    }

    public static void postGpsOff(String apiBase, String token, double lat, double lng) {
        IO.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("lat", lat);
                body.put("lng", lng);
                body.put("address", "GPS turned off");
                body.put("accuracy", JSONObject.NULL);
                Request request = req(apiBase, token, "/api/attendance/gps-off")
                        .post(RequestBody.create(body.toString(), JSON))
                        .build();
                try (Response res = HTTP.newCall(request).execute()) {
                    if (!res.isSuccessful()) Log.w(TAG, "gps-off " + res.code());
                }
            } catch (Exception e) {
                Log.w(TAG, "postGpsOff failed", e);
            }
        });
    }
}

File: ios-native/AAPAttendance/TrackingApi.swift
import Foundation

enum TrackingApi {
    static func postJSON(path: String, body: [String: Any], completion: ((Bool) -> Void)? = nil) {
        let base = SessionStore.apiBase.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard let url = URL(string: base + path), !SessionStore.token.isEmpty else {
            completion?(false)
            return
        }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("Bearer \(SessionStore.token)", forHTTPHeaderField: "Authorization")
        req.setValue("native", forHTTPHeaderField: "X-Client-Source")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try? JSONSerialization.data(withJSONObject: body)
        req.timeoutInterval = 25
        URLSession.shared.dataTask(with: req) { _, res, _ in
            let code = (res as? HTTPURLResponse)?.statusCode ?? 0
            completion?(code >= 200 && code < 300 || code == 429)
        }.resume()
    }

    static func postIntervalSnapshot(slot: Int, lat: Double, lng: Double, completion: ((Bool) -> Void)? = nil) {
        postJSON(path: "/api/attendance/interval-snapshot", body: [
            "slot": slot,
            "lat": lat,
            "lng": lng,
        ], completion: completion)
    }

    static func postHeartbeat(lat: Double, lng: Double) {
        postJSON(path: "/api/attendance/track", body: [
            "points": [] as [Any],
            "heartbeat": ["lat": lat, "lng": lng],
        ])
    }

    static func postTrack(lat: Double, lng: Double, accuracy: Double) {
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        postJSON(path: "/api/attendance/track", body: [
            "points": [[
                "lat": lat,
                "lng": lng,
                "recordedAt": iso.string(from: Date()),
                "accuracy": accuracy,
            ]],
            "heartbeat": ["lat": lat, "lng": lng],
        ])
    }

    static func postGpsOff(lat: Double, lng: Double) {
        postJSON(path: "/api/attendance/gps-off", body: [
            "lat": lat,
            "lng": lng,
            "address": "GPS turned off",
        ])
    }

    static func postSecurityEvent(type: String, action: String, detail: String, lat: Double? = nil, lng: Double? = nil) {
        var body: [String: Any] = [
            "type": type,
            "action": action,
            "detail": detail,
        ]
        if let lat { body["lat"] = lat }
        if let lng { body["lng"] = lng }
        postJSON(path: "/api/attendance/security-event", body: body)
    }
}
